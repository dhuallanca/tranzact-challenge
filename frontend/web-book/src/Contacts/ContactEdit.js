import { useEffect, useState } from 'react';
import axios from '../axios.config';
import { useParams } from "react-router";
import { Alert, Button, Card, Form } from 'react-bootstrap';
import { Link } from 'react-router-dom';
import './Contacts.css';

const ContactEdit = () => {
  const initialContact = {};
  const [contact, setContact] = useState(initialContact);
  const [succeed, setSucceed] = useState(false);
  const params = useParams();
  const contactId = params?.id;

  useEffect(() => {
    getContact(contactId);
  }, [contactId]);

  const getContact = async (id) => {
    const contacts = await axios.get(`/contacts/${id}`);
    setContact(contacts.data);
  };

  const handleOnChange = (field, value) => {
    setContact({
      ...contact,
      [field]: value
    });
  }

  const handleSave = () => {
    if (contact?.contactId) {
      updateContacts(contact.contactId);
      return;
    }
    addContacts();
  }

  const updateContacts = async (id) => {
    const result = await axios.put(`/contacts/${id}`, contact)
      .catch((error) => {
        setSucceed(false);
      });

    if (result?.status === 200) {
      setSucceed(true);
    }
  };

  const addContacts = async () => {
    const result = await axios.post(`/contacts`, contact)
      .catch((error) => {
        setSucceed(false);
      });

    if (result?.status === 200) {
      setSucceed(true);
    }
  };

  if (succeed) {
    return (
      <Alert  variant='success'>
        Contact Successfully Saved!
        <br />
        <hr/>
        <Alert.Link href="/">Contacts List</Alert.Link>
      </Alert>
    )
  }

  return (
    <Card border="secondary">
      <Card.Header>Contact</Card.Header>
      <Card.Body>
        <Form>
        <Form.Group className="mb-3" controlId="formFirstName">
          <Form.Label>First name</Form.Label>
          <Form.Control type="text"
            placeholder="Enter First name"
            value={contact?.firstName}
            onChange={e => handleOnChange('firstName', e.target.value)} />
        </Form.Group>
        <Form.Group className="mb-3" controlId="formLastName">
          <Form.Label>Last Name</Form.Label>
          <Form.Control type="text"
            placeholder="Enter Last name"
            value={contact?.lastName}
            onChange={e => handleOnChange('lastName', e.target.value)}/>
        </Form.Group>
        <Form.Group className="mb-3" controlId="formEmail">
          <Form.Label>Email</Form.Label>
          <Form.Control type="email"
            placeholder="name@domain.com"
            value={contact?.email}
            onChange={e => handleOnChange('email', e.target.value)}/>
        </Form.Group>
        <Form.Group className="mb-3" controlId="formAddress">
          <Form.Label>Address</Form.Label>
          <Form.Control type="text"
            placeholder="Enter Address"
            value={contact?.address}
            onChange={e => handleOnChange('address', e.target.value)}/>
        </Form.Group>
        <Form.Group className="mb-3" controlId="formPhone">
          <Form.Label>Phone number</Form.Label>
          <Form.Control type="text"
            placeholder="Enter Phone number"
            value={contact?.phone}
            onChange={e => handleOnChange('phone', e.target.value)}
          />
        </Form.Group>
        <Button variant="primary" type="button" onClick={handleSave}>
          Save
          </Button>
          <Link to="/"> Back</Link>
      </Form>
      </Card.Body>
    </Card>
  );
}
export default ContactEdit;