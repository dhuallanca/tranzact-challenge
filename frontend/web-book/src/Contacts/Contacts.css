a {
    padding: 1rem;
    float: right;
}

table {
    margin-top: 1rem;
}

thead {
    background-color: darkslategray;
    color: white;
}

tr {
    cursor: pointer;
}

.td-center-align {
    text-align: center;
}