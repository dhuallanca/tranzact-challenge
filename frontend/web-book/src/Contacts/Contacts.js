import React, { useState, useEffect } from 'react';
import { Button, Card, Table } from 'react-bootstrap';
import { Link } from 'react-router-dom';
import { useHistory } from "react-router";
import axios from '../axios.config';
import './Contacts.css';

const Contacts = () => {
  const [contacts, setContacts] = useState([]);

  useEffect(() => {
    getContacts();
  }, []);

  const history = useHistory();

  const getContacts = async () => {
    const contacts = await axios.get('/contacts');
    setContacts(contacts.data);
  };

  const goToUpdate = (id) => {
    history.push(`/Contact/${id}`);
  }

  const handleDelete = async (id, e) => {
    e.stopPropagation();
    const result = await axios.delete(`/contacts/${id}`)
      .catch((error) => {

      });

    if (result?.status === 200) {
      getContacts();
    }
  }

  return (
    <Card border="secondary">
      <Card.Header>
        Contacts
      </Card.Header>
      <Card.Body>
        <Link to="/Contact">Add contact</Link>
        <Table striped bordered hover size="sm">
        <thead>
          <tr>
            <th>First Name</th>
            <th>Last Name</th>
            <th>Email</th>
            <th></th>
            <th></th>
          </tr>
        </thead>
        <tbody>
          {contacts?.map((contact) => (
            <tr key={contact.contactId}>
              <td>{contact.firstName}</td>
              <td>{contact.lastName}</td>
              <td>{contact.email}</td>
              <td className="td-center-align"><Button onClick={() => goToUpdate(contact.contactId)}>
                    Edit
                  </Button>
              </td>
              <td className="td-center-align"><Button variant="danger" onClick={() => handleDelete(contact.contactId)}>
                    Delete
                  </Button>
              </td>
            </tr>
            )
          )}
        </tbody>
      </Table>

      </Card.Body>
    </Card>
  );

}

export default Contacts;