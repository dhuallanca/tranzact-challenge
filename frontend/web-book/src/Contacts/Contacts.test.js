import React from 'react';
import { shallow, mount, render } from 'enzyme';
import toJson from 'enzyme-to-json';
import axios from '../axios.config'
import { BrowserRouter } from 'react-router-dom';
import { act } from "react-dom/test-utils";
import Contacts from './Contacts';

const contactList = [
  {
    contactId: 1,
    firstName: 'Dennis',
    lastName: 'Huallanca',
    email: '',
    address: '',
    phone: ''
  },
  {
    contactId: 2,
    firstName: 'Cesar',
    lastName: 'Gomez',
    email: '',
    address: '',
    phone: ''
  }
]

jest.mock('../axios.config.js', () => ({
  get: jest.fn().mockResolvedValue(),
  create: jest.fn()
}));

describe('Contacts testing', () => {

  let wrapper;

  beforeEach(async () => {
    await act(async () => {
      axios.get.mockResolvedValueOnce({ data: contactList });
      wrapper = mount(<BrowserRouter><Contacts></Contacts></BrowserRouter>);
    });
  });

  it('fetches successfully contacts from API', async () => {
    wrapper.update();
    wrapper.find('button').at(0).simulate('click');
    expect(wrapper.find('button').at(0).text()).toEqual('Edit');
  });

  it('match table', async () => {
    expect(toJson(wrapper.html())).toMatchSnapshot();
  });
})