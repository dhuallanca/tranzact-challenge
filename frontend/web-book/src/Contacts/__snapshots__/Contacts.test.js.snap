// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contacts testing match table 1`] = `undefined`;
